mlir/lib/Dialect/Rock/Transforms/GemmOutputSwizzle.cpp

```diff
@@ -0,0 +1,556 @@
+        op.getExtraIndices(), op.getFeatures(), op.getStoreMethod(),
+        forceUnroll, useIndexDiffs);
+    return success();
+  }
+};
+
+void RockGemmOutputSwizzlePass::runOnOperation() {
+  func::FuncOp func = getOperation();
+  IRRewriter rewriter(func->getContext());
+
+  // Only run this pass on GPU kernel functions.
+  if (!func->hasAttr("kernel"))
+    return;
+
+  // Get total LDS memory allocated
+  int64_t ldsAllocated = getLDSTotalSize(func);
+
+  SmallVector<Operation *, 4> writes;
+  func.walk([&writes, &rewriter,
+             ldsAllocated](ThreadwiseWriteAllOp threadwiseWriteAll) {
+    MemRefType destMemRefType =
+        cast<MemRefType>(threadwiseWriteAll.getDest().getType());
+
+    // process ThreadwiseWriteAllOp that save to global memory
+    if (hasGlobalMemoryAddressSpace(destMemRefType)) {
+      int64_t mPerBlock, nPerBlock;
+      ArrayAttr idToLDS;
+      rewriter.setInsertionPoint(threadwiseWriteAll);
+      std::optional<std::tuple<int64_t, int64_t, ArrayAttr>> maybeBlockInfo =
+          getIdToLDS(threadwiseWriteAll, rewriter);
+      if (!maybeBlockInfo.has_value()) {
+        return;
+      }
+      std::tie(mPerBlock, nPerBlock, idToLDS) = maybeBlockInfo.value();
+
+      Type destType = threadwiseWriteAll.getDest().getType().getElementType();
+      int64_t ldsRequiredBytes = mPerBlock * nPerBlock * getByteWidth(destType);
+
+      // not enough LDS memory
+      if (failed(checkLDSSize(threadwiseWriteAll, ldsRequiredBytes))) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "GemmOutputSwizzle requires too much LDS memory: "
+                   << ldsRequiredBytes << " bytes, skipping pass\n");
+        return;
+      }
+      // heuristic: if we need more LDS, skip this pass
+      if (ldsRequiredBytes > ldsAllocated) {
+        LLVM_DEBUG(
+            llvm::dbgs()
+            << "GemmOutputSwizzle requires more LDS memory, current usage: "
+            << ldsAllocated << " bytes, required: " << ldsRequiredBytes
+            << " bytes, skipping pass\n");
+        return;
+      }
+      // heuristic: check vectorization of iter in the original map
+      ArrayAttr srcTransform = threadwiseWriteAll.getExtraViews();
+      Value matC = threadwiseWriteAll.getDest();
+      Value destView = transform(rewriter, matC, srcTransform);
+      auto destViewType = cast<MemRefType>(destView.getType());
+      auto destElemType = destViewType.getElementType();
+      if (auto elemVecType = dyn_cast<VectorType>(destElemType)) {
+        LLVM_DEBUG(llvm::dbgs() << "ThreadwiseWriteAllOp saves a vector type"
+                                << ", skipping swizzle\n");
+      } else {
+        size_t extraIdxCount = threadwiseWriteAll.getExtraIndices().size();
+        VectorizationResult vectorRes =
+            getMaxVectorization(destView, extraIdxCount);
+        int64_t vectorLenBits =
+            vectorRes.max * destElemType.getIntOrFloatBitWidth();
+        if (vectorLenBits == 128) {
```

Comment:
Yeah, I'm not sure this is what we want for fp16? Maybe `>= 4` here? Worth checking - though I would'n re-tune to test this